### PR TITLE
Fixup nodecpustats testcase

### PR DESCRIPTION
**Pull request by sathnaga**
1. Refactored expected logic to assert from /proc/stat
instead of uptime, which was not right.

2. Removed autotest dependancy for assertion.

Signed-off-by: Satheesh Rajendran <sathnaga@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by sathnaga** [comment]
```
(1/6) guest_import.qemu.qcow2.virtio_scsi.smp2.virtio_net.Fedora.25.ppc64le.powerkvm-qemu.unattended_install.import.import.default_install.aio_native: PASS (40.92 s)
 (2/6) guest_cpu.nodecpusats.qemu.qcow2.virtio_scsi.smp2.virtio_net.Fedora.25.ppc64le.powerkvm-libvirt.virsh.nodecpustats.no_option: PASS (227.01 s)
 (3/6) guest_cpu.nodecpusats.qemu.qcow2.virtio_scsi.smp2.virtio_net.Fedora.25.ppc64le.powerkvm-libvirt.virsh.nodecpustats.unexpect_option.invalid_option: PASS (5.04 s)
 (4/6) guest_cpu.nodecpusats.qemu.qcow2.virtio_scsi.smp2.virtio_net.Fedora.25.ppc64le.powerkvm-libvirt.virsh.nodecpustats.unexpect_option.invalid_cpuNum: PASS (5.23 s)
 (5/6) guest_cpu.nodecpusats.qemu.qcow2.virtio_scsi.smp2.virtio_net.Fedora.25.ppc64le.powerkvm-libvirt.virsh.nodecpustats.with_libvirtd_stop: PASS (6.67 s)
 (6/6) guest_remove.qemu.qcow2.virtio_scsi.smp2.virtio_net.Fedora.25.ppc64le.powerkvm-libvirt.remove_guest.without_disk: PASS (14.18 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
TESTS TIME : 299.05 s

```

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py [hunk 1]
```diff
@@ -1,6 +1,5 @@
 import re
 
-from autotest.client.shared import error
 from autotest.client import utils
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: autotest dependency can be removed,
`utils_misc.get_cpu_processors()` from virttest can be replaced for `utils.cpu_online_map() `from autotest

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py [hunk 2]
```diff
@@ -23,57 +22,94 @@ def run(test, params, env):
...
+            test.fail("Commands 'virsh nodecpustats' not succeeded"
+                      " as total time: %f is more"
+                      " than proc/stat: %f" % (total, uptime))
         return True
 
     def virsh_check_nodecpustats_percentage(actual_per):
         """
         Check the actual nodecpustats percentage adds up to 100%
+
+        param actual_stats: Actual cpu stats percentage
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: `:param actual_per: Actual cpu stats percentage`

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py [hunk 3]
```diff
@@ -23,57 +22,94 @@ def run(test, params, env):
...
+            test.fail("Commands 'virsh nodecpustats' not succeeded"
+                      " as the total percentage value: %d"
+                      " is not equal 100" % total)
 
     def parse_output(output):
         """
         To get the output parsed into a dictionary
-        :param virsh command output
 
+        :param virsh command output
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: `:param output: virsh command output`

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py [hunk 4]
```diff
@@ -99,8 +135,8 @@ def parse_output(output):
     def parse_percentage_output(output):
         """
         To get the output parsed into a dictionary
-        :param virsh command output
 
+        :param virsh command output
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: same here

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py [hunk 5]
```diff
@@ -23,57 +22,94 @@ def run(test, params, env):
...
+        stats['user'] = int(cpu_stat[idx][0]) + int(cpu_stat[idx][1])
+        stats['system'] = int(cpu_stat[idx][2]) + int(cpu_stat[idx][5]) + int(cpu_stat[idx][6])
+        stats['idle'] = int(cpu_stat[idx][3])
+        stats['iowait'] = int(cpu_stat[idx][4])
+        stats['total'] = stats['user'] + stats['system'] + stats['idle'] + stats['iowait']
+        return stats
+
+    def virsh_check_nodecpustats_percpu(actual_stats, cpu):
         """
         Check the acual nodecpustats output value
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: %s/acual/actual

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py [hunk 6]
```diff
@@ -23,57 +22,94 @@ def run(test, params, env):
...
-        total time <= system uptime
+        total time <= total stat from proc
+
+        param actual_stats: Actual cpu stats
+        param cpu: cpu index
+
+        return: True if matches, else failout
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]:
```
:param actual_stats: Actual cpu stats
:param cpu: cpu index

:return: True if matches, else failout
```

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py (hunk 1 above)
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: Sure

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py (hunk 5 above)
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: :-), done

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py (hunk 6 above)
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: Sure

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py (hunk 2 above)
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: Sure

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py (hunk 3 above)
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: Sure

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py (hunk 4 above)
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: Sure

**Comment by sathnaga** [comment]
@balamuruhans Thanks for the review, addressed the comments :-)

**Comment by sathnaga** [comment]
@balamuruhans Thanks for the review, addressed the comments :-)

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py [hunk 7]
```diff
@@ -23,57 +21,94 @@ def run(test, params, env):
...
+            test.fail("Commands 'virsh nodecpustats' not succeeded"
+                      " as total time: %f is more"
+                      " than proc/stat: %f" % (total, uptime))
         return True
 
     def virsh_check_nodecpustats_percentage(actual_per):
         """
         Check the actual nodecpustats percentage adds up to 100%
+
+        :param actual_stats: Actual cpu stats percentage
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: %s/actual_stats/actual_per

**Comment by sathnaga** [comment]
@balamuruhans Done, my bad, Missed Saving.

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py [hunk 8]
```diff
@@ -23,57 +21,94 @@ def run(test, params, env):
...
+        :return: dict of cpu stats
         """
-        Check the acual nodecpustats output value
-        total time <= system uptime
+        stats = {}
+        cpu_stat = []
+        fl = open("/proc/stat", "r")
+        for line in fl.readlines():
+            if line.startswith("cpu"):
+                cpu_stat.append(line.strip().split(" ")[1:])
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: will API `get_load_per_cpu()` from virttest/staging/utils_cgroup.py helps here

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py [hunk 9]
```diff
@@ -23,57 +21,94 @@ def run(test, params, env):
...
+
+        expected = get_expected_stat(cpu)
+        if not total <= expected['total']:
+            test.fail("Commands 'virsh nodecpustats' not succeeded"
+                      " as total time: %f is more"
+                      " than proc/stat: %f" % (total, uptime))
         return True
 
-    def virsh_check_nodecpustats(actual_stats, cpu_count):
+    def virsh_check_nodecpustats(actual_stats):
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: `virsh_check_nodecpustats_percpu()` and `virsh_check_nodecpustats()` are almost similar, let us have single function `virsh_check_nodecpustats(actual_stats, cpu=None)` and call `get_expected_stat(cpu=cpu) `

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py (hunk 8 above)
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: @balamuruhans partially yes, but this groups the values logically which is useful, will think about keeping it in some common place later.

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py (hunk 9 above)
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: @balamuruhans I did not refactor the methods which was there already there, will take it up later in separate PR.

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py [hunk 10]
```diff
@@ -23,57 +21,94 @@ def run(test, params, env):
...
+        Parse cpu stats from /proc/stat
+
+        :param cpu: cpu index, None for total cpu stat
+        :return: dict of cpu stats
         """
-        Check the acual nodecpustats output value
-        total time <= system uptime
+        stats = {}
+        cpu_stat = []
+        fl = open("/proc/stat", "r")
```
Comment (will-Do) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: always close the fd later

Comment (will-Do) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: And other parts looks good, thanks @sathnaga

Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: @will-Do done :-), Thanks

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py (hunk 8 above)
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: fine.

File: libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py (hunk 9 above)
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/host/virsh_nodecpustats.py]: Sure.

**Comment by balamuruhans** [comment]
Thanks @sathnaga @will-Do

**Pull request merged**